server/src/main/java/org/opensearch/index/codec/CodecService.java

```diff
@@ -58,23 +61,26 @@ public class CodecService {
 
     public static final String DEFAULT_CODEC = "default";
     public static final String BEST_COMPRESSION_CODEC = "best_compression";
-    /** the raw unfiltered lucene default. useful for testing */
+    /**
+     * the raw unfiltered lucene default. useful for testing
+     */
     public static final String LUCENE_DEFAULT_CODEC = "lucene_default";
     public static final String ZSTD_CODEC = "zstd";
     public static final String ZSTD_NO_DICT_CODEC = "zstd_no_dict";
 
-    public CodecService(@Nullable MapperService mapperService, Logger logger) {
+    public CodecService(@Nullable MapperService mapperService, IndexSettings indexSettings, Logger logger) {
```

Comment:
Changing the constructor signature may break the plugins which may be initializing the CodecService (if any) outside of the OpenSearch repo. Will we need to mark the current constructor as deprecated and introduce a new constructor taking the required arguments?

Reply:
So we have had many discussions regarding breaking and non-breaking changes, the take away is this class is marked as internal API:

```
* @opensearch.internal
``` 

And for internal API, there should be a freedom in making changes like that (that would not be the case if e changed the `CodecService` wiring in  `EnginePlugin` fe).

Reply:
The current change may break builds for k-NN plugin: https://github.com/opensearch-project/k-NN/blob/d55eccb4c5150d244ffc676975a6a39685688204/src/main/java/org/opensearch/knn/index/codec/KNNCodecService.java#L21

While the class is marked internal, the way EnginePlugin is designed is that a plugin extending this may need to instantiate CodecService to provide custom codecs. Would like to understand our take here and if we should notify k-NN repo owners if we go ahead with the change.

Reply:
Well, the plugin owners (hopefully) are building against 3.0.0-SNAPSHOT, so they would see the change is needed, that applies to all plugins, no k-NN only.

Reply:
@reta in general we can introduce breaking changes in next major version which is going to be 3.0.0 even in the EnginePlugin right? For my understanding, the classes which are returned as custom implementation from plugin interfaces, those shouldn't be marked as Internal as that is part of the contract and hence can't have breaking changes in same major version.

Reply:
> @reta in general we can introduce breaking changes in next major version which is going to be 3.0.0 even in the EnginePlugin right?

Correct

> For my understanding, the classes which are returned as custom implementation (e.g. CodecService here) from plugin interfaces, those shouldn't be marked as Internal as that is part of the contract and hence can't have breaking changes in same major version.

You are very right, sadly this is not the case, but I am 100% agree on this statement. Since 3.0.0 is a major version - the breaking changes like that are well justified.

Reply:
@reta @shwetathareja with an intention to not backport this change, are we good to merge this in the 3.0.0?

Reply:
I think we're good even if we backport, see @reta's comment above. 

@mgodwan you're good with this? (resolve?)

Reply:
@dblock Yes, I am aligned on moving ahead with this.

Reply:
For 2.x backport, we would need to have a backward compatible change (with the overloaded constructor)